### ISSUE TITLE
Add example for simple upgradeable contract

### ISSUE THREAD

**Issue by cmichi** [paritytech/ink#698]
We should add an example for a simple upgradeable contract, which uses 3-layers:
* A `delegate` contract which acts as a proxy and contains the address of the `logic` contract to call. The `logic` contract address can be updated only by an authorized address.
* A `logic` contract which contains the business logic.
* A `data`/`model` contract which exposes access to the data storage.

**Comment by HCastano** [comment]
@cmichi Hey, are you still working on this? If not, can I give it a go?

**Comment by cmichi** [comment]
@HCastano Feel free to give it a go, but the whole thing is way more complicated unfortunately, which is why I've put it on hold in April.

So the short version is that the issue needs a couple things to be implemented before it can be started. This is because the `contracts` pallet (and ink!) are currently lacking the means of forwarding a call. This is necessary for implementing the proxy pattern, where the proxy contract forwards the call to another contract. At the time we discussed this and came up with https://github.com/paritytech/ink/issues/739 to allow forwarding calls. So that issue has to be implemented before this one here can be started, since we can't have a contract that forwards calls otherwise.

But the story doesn't stop there, this issue https://github.com/paritytech/ink/issues/739 is blocked by https://github.com/paritytech/substrate/pull/8909 unfortunately though. But that one is about to be merged.

**Comment by HCastano** [comment]
@cmichi why'd you close this. We're close to actually having the primitives (`delegate_call` and `set_code`) to do this
